core/src/main/java/io/questdb/griffin/engine/table/SequentialRecordCursorFactory.java

```diff
@@ -0,0 +1,339 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2023 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *2
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.table;
+
+import io.questdb.cairo.*;
+import io.questdb.cairo.sql.*;
+import io.questdb.cairo.sql.Record;
+import io.questdb.griffin.AnyRecordMetadata;
+import io.questdb.griffin.PlanSink;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.std.*;
+import io.questdb.std.str.CharSink;
+
+import java.io.Closeable;
+import java.io.IOException;
+
+/**
+ * Returns rows from current record in order of cursors list :
+ * - first fetches and returns all records from first cursor
+ * - then from second cursor, third, ...
+ * until all cursors are exhausted .
+ */
+public abstract class SequentialRecordCursorFactory<T extends RecordCursorFactory> implements RecordCursorFactory {
+    private final SequentialRecordCursor cursor;
+    private final ObjList<T> cursorFactories;
+    private final ObjList<RecordCursor> cursors;
+
+    public SequentialRecordCursorFactory() {
+        cursorFactories = new ObjList<>();
+        cursors = new ObjList<>();
+        cursor = new SequentialRecordCursor();
+    }
+
+    public abstract RecordMetadata getMetadata();
+
+    @Override
+    public boolean recordCursorSupportsRandomAccess() {
+        return false;
+    }
+
+
+    public abstract void initFactories(SqlExecutionContext executionContext, ObjList<T> factoriesBucket);
+
+    @Override
+    public RecordCursor getCursor(SqlExecutionContext executionContext) throws SqlException {
+        for (int i = 0; i < cursorFactories.size(); i++) {
+            T factory = cursorFactories.getQuick(i);
+            factory.close();
+        }
+        this.cursorFactories.clear();
+
+        initFactories(executionContext, this.cursorFactories);
+
```

Comment:
Closing factories should close associated cursors so closing them here isn't necessary .